tlogclient (in tlogStorage) fails at master (c56f16a): max send retry reached

logs:

```
t=2017-06-11T13:29:15+0000 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:36976" module=global caller=redis.go:47
t=2017-06-11T13:29:15+0000 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:41888" module=global caller=redis.go:47
t=2017-06-11T13:29:15+0000 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:42736" module=global caller=redis.go:47
t=2017-06-11T13:29:15+0000 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:37965" module=global caller=redis.go:47
t=2017-06-11T13:29:15+0000 lvl=info msg="Started listening on local address 127.0.0.1:39693" module=global caller=server.go:55
t=2017-06-11T13:29:15+0000 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:36231" module=global caller=redis.go:47
t=2017-06-11T13:29:15+0000 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:42928" module=global caller=redis.go:47
t=2017-06-11T13:29:15+0000 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:37918" module=global caller=redis.go:47
t=2017-06-11T13:29:15+0000 lvl=dbug msg="received handshake request from incoming connection" module=global caller=server.go:140
t=2017-06-11T13:29:15+0000 lvl=dbug msg="incoming connection checks out with version: 1.1.0-alpha" module=global caller=server.go:150
t=2017-06-11T13:29:15+0000 lvl=dbug msg="create vdisk with expectedSequence:0" module=global caller=vdisk.go:147
t=2017-06-11T13:29:15+0000 lvl=info msg="vdisk a reset first sequence to 0" module=global caller=vdisk.go:222
t=2017-06-11T13:29:15+0000 lvl=dbug msg="handshake phase successfully completed" module=global caller=server.go:175
t=2017-06-11T13:29:15+0000 lvl=dbug msg="replying handshake with status: OK" module=global caller=server.go:202
t=2017-06-11T13:29:15+0000 lvl=dbug msg="vdisk a's tlog server has received force flush message" module=global caller=tlog.go:284
t=2017-06-11T13:29:15+0000 lvl=dbug msg="vdisk a's tlog server has received force flush message" module=global caller=tlog.go:284
panic: tlogStorage couldn't send block 0 (seq: 0): max send retry reached
goroutine 20097 [running]:
github.com/zero-os/0-Disk/nbdserver/ardb.(*tlogStorage).transactionSender(0xc4204087e0, 0x118aba0, 0xc4208186c0)
	/home/travis/gopath/src/github.com/zero-os/0-Disk/nbdserver/ardb/tlog.go:322 +0x2b9
created by github.com/zero-os/0-Disk/nbdserver/ardb.(*tlogStorage).GoBackground
	/home/travis/gopath/src/github.com/zero-os/0-Disk/nbdserver/ardb/tlog.go:260 +0xc2
FAIL	github.com/zero-os/0-Disk/nbdserver/ardb	25.680s
```

Job: https://travis-ci.org/zero-os/0-Disk/jobs/241728348

```
t=2017-06-11T13:29:15+0000 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:36976" module=global caller=redis.go:47
t=2017-06-11T13:29:15+0000 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:41888" module=global caller=redis.go:47
t=2017-06-11T13:29:15+0000 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:42736" module=global caller=redis.go:47
t=2017-06-11T13:29:15+0000 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:37965" module=global caller=redis.go:47
t=2017-06-11T13:29:15+0000 lvl=info msg="Started listening on local address 127.0.0.1:39693" module=global caller=server.go:55
t=2017-06-11T13:29:15+0000 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:36231" module=global caller=redis.go:47
t=2017-06-11T13:29:15+0000 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:42928" module=global caller=redis.go:47
t=2017-06-11T13:29:15+0000 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:37918" module=global caller=redis.go:47
t=2017-06-11T13:29:15+0000 lvl=dbug msg="received handshake request from incoming connection" module=global caller=server.go:140
t=2017-06-11T13:29:15+0000 lvl=dbug msg="incoming connection checks out with version: 1.1.0-alpha" module=global caller=server.go:150
t=2017-06-11T13:29:15+0000 lvl=dbug msg="create vdisk with expectedSequence:0" module=global caller=vdisk.go:147
t=2017-06-11T13:29:15+0000 lvl=info msg="vdisk a reset first sequence to 0" module=global caller=vdisk.go:222
t=2017-06-11T13:29:15+0000 lvl=dbug msg="handshake phase successfully completed" module=global caller=server.go:175
t=2017-06-11T13:29:15+0000 lvl=dbug msg="replying handshake with status: OK" module=global caller=server.go:202
t=2017-06-11T13:29:15+0000 lvl=dbug msg="vdisk a's tlog server has received force flush message" module=global caller=tlog.go:284
t=2017-06-11T13:29:15+0000 lvl=dbug msg="vdisk a's tlog server has received force flush message" module=global caller=tlog.go:284
panic: tlogStorage couldn't send block 0 (seq: 0): max send retry reached

goroutine 20097 [running]:
github.com/zero-os/0-Disk/nbdserver/ardb.(*tlogStorage).transactionSender(0xc4204087e0, 0x118aba0, 0xc4208186c0)
	/home/travis/gopath/src/github.com/zero-os/0-Disk/nbdserver/ardb/tlog.go:322 +0x2b9
created by github.com/zero-os/0-Disk/nbdserver/ardb.(*tlogStorage).GoBackground
	/home/travis/gopath/src/github.com/zero-os/0-Disk/nbdserver/ardb/tlog.go:260 +0xc2
FAIL	github.com/zero-os/0-Disk/nbdserver/ardb	25.680s
```

Job: https://travis-ci.org/zero-os/0-Disk/jobs/241728347

--- a/nbdserver/ardb/tlog.go
+++ b/nbdserver/ardb/tlog.go
@@ -315,6 +315,9 @@ func (tls *tlogStorage) transactionSender(ctx context.Context) {
 				transaction.Size,
 			)
 			if err != nil {
+				if err == tlogclient.ErrClientClosed {
+					return
+				}
 				panic(fmt.Errorf(
 					"tlogStorage couldn't send block %d (seq: %d): %s",
 					transaction.Offset/uint64(tls.blockSize),

--- a/nbdserver/ardb/tlog.go
+++ b/nbdserver/ardb/tlog.go
@@ -248,6 +248,7 @@ func (tls *tlogStorage) Close() (err error) {
 // GoBackground implements backendStorage.GoBackground
 func (tls *tlogStorage) GoBackground(ctx context.Context) {
 	defer func() {
+		log.Infof("GoBackground exited for vdisk: %v", tls.vdiskID)
 		err := tls.tlog.Close()
 		if err != nil {
 			log.Info("error while closing tlog client: ", err)

--- a/tlog/tlogclient/client.go
+++ b/tlog/tlogclient/client.go
@@ -372,34 +372,32 @@ func (c *Client) sendReconnect(sender func() (interface{}, error)) (interface{},
 	var err error
 	var ret interface{}
 
-	okToSend := true
-
 	for i := 0; i < sendRetryNum+1; i++ {
-		if okToSend {
-			ret, err = sender()
-			if err == nil {
-				return ret, nil
-			}
+		if err != nil {
+			closedTime := time.Now()
 
-			if _, ok := err.(net.Error); !ok {
-				return nil, err // no need to rety if it is not network error.
+			// First sleep = 0 second,
+			// so we don't need to sleep in case of simple closed connection.
+			// We sleep in next iteration because there might be something error in
+			// the network connection or the tlog server that need time to be recovered.
+			time.Sleep(time.Duration(i-1) * sendSleepTime)
+
+			if err = c.reconnect(closedTime); err != nil {
+				log.Infof("tlog client : reconnect from send attemp(%v) failed:%v", i, err)
+				continue
 			}
 		}
 
-		closedTime := time.Now()
-
-		// First sleep = 0 second,
-		// so we don't need to sleep in case of simple closed connection.
-		// We sleep in next iteration because there might be something error in
-		// the network connection or the tlog server that need time to be recovered.
-		time.Sleep(time.Duration(i) * sendSleepTime)
-
-		if err = c.reconnect(closedTime); err != nil {
-			log.Infof("tlog client : reconnect from send attemp(%v) failed:%v", i, err)
-			okToSend = false
-		} else {
-			okToSend = true
+		ret, err = sender()
+		if err == nil {
+			return ret, nil
 		}
+
+		if _, ok := err.(net.Error); !ok {
+			return nil, err // no need to rety if it is not network error.
+		}
+		log.Errorf("tlogclient failed to send: %v", err)
+
 	}
 	return nil, errMaxSendRetry
 }

--- a/tlog/tlogclient/client.go
+++ b/tlog/tlogclient/client.go
@@ -25,7 +25,8 @@ const (
 )
 
 var (
-	errMaxSendRetry = errors.New("max send retry reached")
+	errMaxSendRetry     = errors.New("max send retry reached")
+	errReconnectStopped = errors.New("reconnect stopped client")
 )
 
 // Response defines a response from tlog server
@@ -92,7 +93,7 @@ func New(addr, vdiskID string, firstSequence uint64, resetFirstSeq bool) (*Clien
 // it must be called under wLock
 func (c *Client) reconnect(closedTime time.Time) (err error) {
 	if c.stopped {
-		return nil
+		return errReconnectStopped
 	}
 	if c.lastConnected.After(closedTime) { // another goroutine made the connection work again
 		return

--- a/tlog/tlogclient/client.go
+++ b/tlog/tlogclient/client.go
@@ -25,8 +25,11 @@ const (
 )
 
 var (
-	errMaxSendRetry     = errors.New("max send retry reached")
-	errReconnectStopped = errors.New("reconnect stopped client")
+	errMaxSendRetry = errors.New("max send retry reached")
+
+	// ErrClientClosed returned when client do something when
+	// it already closed
+	ErrClientClosed = errors.New("client already closed")
 )
 
 // Response defines a response from tlog server
@@ -93,8 +96,9 @@ func New(addr, vdiskID string, firstSequence uint64, resetFirstSeq bool) (*Clien
 // it must be called under wLock
 func (c *Client) reconnect(closedTime time.Time) (err error) {
 	if c.stopped {
-		return errReconnectStopped
+		return ErrClientClosed
 	}
+
 	if c.lastConnected.After(closedTime) { // another goroutine made the connection work again
 		return
 	}
@@ -375,6 +379,10 @@ func (c *Client) sendReconnect(sender func() (interface{}, error)) (interface{},
 
 	for i := 0; i < sendRetryNum+1; i++ {
 		if err != nil {
+			if _, ok := err.(net.Error); !ok {
+				return nil, err // no need to rety if it is not network error.
+			}
+
 			closedTime := time.Now()
 
 			// First sleep = 0 second,
@@ -392,10 +400,6 @@ func (c *Client) sendReconnect(sender func() (interface{}, error)) (interface{},
 		ret, err = sender()
 		if err == nil {
 			return ret, nil
-		}
-
-		if _, ok := err.(net.Error); !ok {
-			return nil, err // no need to rety if it is not network error.
 		}
 		log.Errorf("tlogclient failed to send: %v", err)
 

